examples/small/lazyeval.makam

```diff
@@ -0,0 +1,196 @@
+(* An implementation of the standard Launchbury semantics
+   for call-by-need evaluation, using an explicit heap. *)
+
+expr : type.
+var : type.
+
+lam : (var -> expr) -> expr.
+
+(* Arguments to applications need to be variables *)
+app : expr -> var -> expr.
+
+var : var -> expr.
+
+let : bindmany var (list expr * expr) -> expr.
+
+intconst : int -> expr.
+
+binop : (int -> int -> int -> prop) -> expr -> expr -> expr.
+
+(* now let's model heaps. In this semantics, heaps are
+   variable binding - expression definition pairs. In Makam,
+   we will model them as a number of bindings followed by
+   their definitions in the same order.
+*)
+
+heap : type -> type.
+heap : bindmany var (list expr * A) -> heap A.
+
+%extend heap.
+
+open : heap A -> (list var -> list expr -> A -> prop) -> prop.
+
+open (heap XS_DefsBody) P :-
+  bindmany.open XS_DefsBody (pfun XS (Defs, Body) =>
+    P XS Defs Body).
+
+apply : heap A -> list var -> list expr -> A -> prop.
+apply (heap (body (Defs, Body))) [] Defs Body.
+apply (heap (bind Name Rest)) (X :: XS) Defs Body :-
+  unless (nameofvar X Name) (success),
+  apply (heap (Rest X)) XS Defs Body.
+
+apply_needed : heap A -> list var -> A -> prop.
+apply_needed (heap (body (_, Body))) _ Body.
+apply_needed (heap (bind Name Rest)) (X :: XS) Body :-
+  unless (nameofvar X Name) (success),
```

Comment:
That part is mostly for "copying" the names of variables from one heap to a new one.
So, say we have a concrete heap `H` that we have `open`ed; from that we get the list of variables `XS` that we apply here; and we want to create a new heap `H'`. Within `apply_needed H' XS ..`, this would make the bindings of the new heap associated with the same concrete names as `XS`, rather than leaving them undefined. (See, for example, what happens if you comment out these calls and run one of the example queries.)

Maybe a note here, in case you haven't seen the hybrid treatment of variables in `bindmany` and friends: each variable is both a normal HOAS binding (that captures the binding part of the variable), but is also associated with a concrete string, which is an optional concrete name for the variable and used, say, for pretty-printing. That concrete string can be alpha-renamed if needed.

Here I'm trying to preserve the concrete names; that said, there are cases where those concrete names will need to be alpha-renamed because they're duplicated, e.g. in the presence of recursion.

Reply:
Thanks!